Support accessor set input

averissimo opened the pull request:
Related issues
- Closes #43.

## Changes description (include a screenshot if applicable!)

- Support accessor as well as an index in `setInput`
- Adds unit tests in R

![Peek 2023-01-20 18-55](https://user-images.githubusercontent.com/211358/213782493-79fb3d6d-9b16-4996-b2e5-dba3d4f37020.gif)

## Definition of Done checklist
- [x] :spiral_notepad: README, other documentation and code comments that we have is updated with all information related to the change.
- [x] Unit tests added for all new or changed logic.


- [ ] End-to-end tests added for all new or changed logic.

kamilzyla commented on R/react.R:
```diff
@@ -161,18 +161,82 @@ triggerEvent <- function(inputId) {
+methods::setMethod(
+  "setInput",
+  signature(inputId = "character", argIdx = "numeric"),
+  function(inputId, argIdx = 1) {
```
Perhaps we shall change to 0-based indexing? It might be confusing that `setInput(id, 1)` and `setInput(id, "[0]")` are equivalent and `setInput()` can be seen as a shorthand for `JS()`, so 0-based indexing might be more natural actually. WDYT?

kamilzyla commented on R/react.R:
```diff
@@ -161,18 +161,82 @@ triggerEvent <- function(inputId) {
+#' setInput("some_id")
+methods::setGeneric(
+  "setInput",
+  function(inputId, argIdx = 1) {
```
I'm no expert on S4, but this default value seems not to be used?
```suggestion
  function(inputId, argIdx) {
```

kamilzyla commented on tests/testthat/test-setInput.R:
```diff
@@ -0,0 +1,37 @@
+library(testthat)
```
Kudos for adding the firsts R tests! :grin: Let's remove `test-all.R`. We can also get rid of `mockery` from `DESCRIPTION`.

kamilzyla commented on js/webpack.config.babel.js:
```diff
@@ -5,6 +5,7 @@ export default {
   output: {
     path: join(__dirname, '..', 'inst', 'www', 'shiny.react'),
     filename: 'shiny-react.js',
+    hashFunction: "sha256",
```
Can you tell me more, why is this necessary? Ideally we'd fix your setup so it works without this line and perhaps update docs. If that's unpractical or we figure we should actually explicitly set `hashFunction`, please use single quotes (consistency) and add a comment (why such setting).

averissimo replied:
From my understanding openssl removed support for the hashing algorithm _(md4)_ that older versions of webpack default to. In particular, with the one that's currently used by `shiny.react`.

So when running webpack it just fails.

![image](https://user-images.githubusercontent.com/211358/214328942-4dc15f1d-cab6-4b3b-a7c4-8a86732502fc.png)

(this is what happens when I run `yarn && yarn webpack` in master)

See [here](https://stackoverflow.com/a/69476335/302485)

If we update webpack to v5.54.0+ this problem goes away and we can remove this output option

**I just didn't want to update node packages along with this PR, but if you're ok with it I can.**

averissimo replied:
I've updated the webpack and this solves the problem

kamilzyla commented on js/src/react/mapReactData.js:
```diff
@@ -85,9 +85,10 @@ dataMappers.element = ({ module, name, props: propsData }) => {
+dataMappers.input = ({ id, argIdx, accessor }) => (
   (...args) => {
-    const value = argIdx === null ? true : args[argIdx];
+    let value = argIdx === null || argIdx === undefined ? true : args[argIdx];
+    value = accessor === null || accessor === undefined ? value : eval(`${args}accessor`); // eslint-disable-line no-eval
```
I find this hard to read. Let's have a separate data mapper for `triggerEvent()` to avoid `=== null` checks. Then consider a single `accessor` argument and check its type, or perhaps have separate data mappers for `argIdx` and `accessor` cases.

averissimo replied:
I preferred using a more explicit if conditional, but that was giving me a weird issue, where `args` became `undefined`.

I'll improve on this

averissimo replied:
I've unified the solution, so that `jsAccessor = 1` resolves to "jsAccessor[1]" in R. This keeps the code cleaner with a single point of maintenance.

`triggerEvent` has its own dataMapper that is straightforward

kamilzyla reviewed (changes requested):
Very neat PR!

averissimo commented on R/react.R:
```diff
@@ -161,18 +161,82 @@ triggerEvent <- function(inputId) {
 #'
 #' @param inputId 'Shiny' input ID to set the value on.
-#' @param argIdx Index of the argument to use as value.
+#' @param argIdx Index (numeric) or accessor (string) of the argument to use as value.
```
I kept the same argument name due to backwards compatibility for the cases where it was called explicitly `setInput('some_id', argIdx = 1)`

If you think we can move past the name, then I agree with a name change to accessor. 

I'll add the extra examples

averissimo commented on R/react.R:
```diff
@@ -161,18 +161,82 @@ triggerEvent <- function(inputId) {
+#'
+#' @description  Prints a description of valid parameters for the `setInput`
+#' function
+setInputUsage <- function() {
```
Will do!

averissimo commented:
I believe I resolved all your comments @kamilzyla 

- removed setInputUsage
- updated webpack so that there's no hashing option to maintain
- rename argIdx to jsAccessor
- break triggerEvent to its own dataMapper called `dataMappers.event`

Please take a look at the changes I did to `setInput` as it seemed redundant to have code in R and JS handling the index vs. accessor

I'm a bit ashamed of the time it took to figure out a bug I was having with `(...args)` to `arguments`

kamilzyla commented on R/react.R:
```diff
@@ -157,22 +157,89 @@ JS <- function(...) { # nolint
+methods::setMethod(
+  "setInput",
+  signature(inputId = "character", jsAccessor = "numeric"),
+  function(inputId, jsAccessor = 1) {
```
Is this default value used in any way? If it is, make it a 0 :wink:
```suggestion
  function(inputId, jsAccessor) {
```

kamilzyla commented on R/react.R:
```diff
@@ -157,22 +157,89 @@ JS <- function(...) { # nolint
+methods::setMethod(
+  "setInput",
+  signature(inputId = "character", jsAccessor = "character"),
+  function(inputId, jsAccessor = NULL) {
```
Is this default value used in any way?
```suggestion
  function(inputId, jsAccessor) {
```

kamilzyla commented on R/react.R:
```diff
@@ -157,22 +157,89 @@ JS <- function(...) { # nolint
-#' Set input
+#' @title Set input
 #'
-#' Creates a handler which can be used for `onChange` and similar props of 'React' components
-#' to set the value of a 'Shiny' input to one of the arguments passed to the handler.
+#' @description Creates a handler which can be used for `onChange` and similar
```
I'd suggest to drop `@title` and `@description`:
1. Consistency: that's how the rest of our docs are written.
2. The first paragraph will become the `@description` and the remaining ones (starting with "The argument...") will become `@details` which makes sense in the rendered docs.

averissimo replied:
Done!

kamilzyla commented on js/src/react/mapReactData.js:
```diff
@@ -82,12 +82,25 @@ dataMappers.element = ({ module, name, props: propsData }) => {
+);
+
+// Used to implement `setInput()` R function.
+dataMappers.input = ({ id, jsAccessor }) => (
+  () => { // used to be (...args) , see below
+    //
+    let value = true;
+    if (jsAccessor !== undefined) {
+      // Needs to use arguments inside eval string, otherwise webpack
+      // won't translate args => arguments
+      value = eval(`arguments${jsAccessor}`); // eslint-disable-line no-eval
+    }
     Shiny.setInputValue(id, value, { priority: 'event' });
   }
 );
```
WDYT about this implementation? I dropped the `jsAccessor !== undefined` check - when could it fail?
```js
dataMappers.input = ({ id, jsAccessor }) => {
  const getValue = eval(`(...args) => (args${jsAccessor})`); // eslint-disable-line no-eval
  return (...args) => {
    Shiny.setInputValue(id, getValue(args), { priority: 'event' });
  }
}
```

kamilzyla replied:
Note: There was a bug in my suggested code above (`...args` instead of `args` in the `eval()`). Now fixed.

averissimo replied:
Very clean!!

I like it a lot and it avoids having to use `arguments` directly as it's no longer inside of a string

I like to safeguard against some stupid bugs, but here it's really not necessary

kamilzyla reviewed (changes requested):
Great job! Have a bunch more comments, mostly nitpicks about documentation :wink:

averissimo commented:
I dropped the ball on the documentation. Thank you for catching it!

kamilzyla reviewed (approved):
LGTM :+1: Great job and thanks for your patience :smiley: :tada: Merging with a tiny docs change of mine.
